pvlib/tools.py

```diff
@@ -425,3 +426,58 @@ def _golden_sect_DataFrame(params, VL, VH, func):
             raise Exception("EXCEPTION:iterations exceeded maximum (50)")
 
     return func(df, 'V1'), df['V1']
+
+
+def enforce_numpy_arrays(f):
+    """ This decorator function will convert any inputted pandas dataframe or
+    series to a numpy array, and make sure to convert outtputed numpy arrays
+    to pandas series if any input was a pandas structure """
+
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+        # Convert all inputs to numpy arrays if pandas series or dataframe
+        is_return_type_pandas = False
+
+        new_args = []
+        for arg in args:
+            if isinstance(arg, pd.Series):
+                is_return_type_pandas = True
+                new_arg = arg.values
+                new_args.append(new_arg)
+            elif isinstance(arg, pd.DataFrame):
+                is_return_type_pandas = True
+                new_arg = arg.values.ravel()  # make sure that 1D
```

Comment:
I think it's better to raise an exception than allow a `DataFrame --> 1D array --> Series` pattern.

Reply:
Thanks, I removed the decorator function and implemented checks in the pvfactors function